### ISSUE TITLE
create_distributed_table as non-schema-owner.

### ISSUE THREAD

**Issue by Rakshitha-BR** [citusdata/citus#2575]
Whenever we try to create distributed table as non schema owner with create privileges, facing the below error. Can you please check on it.

 

ERROR:
```
fpaprd=> SELECT create_distributed_table('a3.test1', 'id');
ERROR:  must be member of role "postgres"
CONTEXT:  while executing command on finpg-rcdn-prd-05:5432
```

Steps to reproduce:

>psql -U postgres -d fpaprd

```sql
create role a3 with login encrypted password 'abcd123';
SELECT run_command_on_workers($cmd$create role a3 with login encrypted password 'abcd123';$cmd$);

create schema a3;
SELECT run_command_on_workers($cmd$create schema a3;$cmd$);

grant connect on database fpaprd to a3;
SELECT run_command_on_workers($cmd$grant connect on database fpaprd to a3;$cmd$);

grant create on database fpaprd to a3;
SELECT run_command_on_workers($cmd$grant create on database fpaprd to a3;$cmd$);

grant usage on schema a3 to a3;
SELECT run_command_on_workers($cmd$grant usage on schema a3 to a3;$cmd$);

grant ALL PRIVILEGES on SCHEMA a3 to a3;
SELECT run_command_on_workers($cmd$grant ALL PRIVILEGES on SCHEMA a3 to a3;$cmd$);

grant create on schema a3 to a3;
SELECT run_command_on_workers($cmd$grant create on schema a3 to a3;$cmd$);
```

>psql -U a3 -d fpaprd

```sql
CREATE TABLE IF NOT EXISTS a3.test1(id bigint, name text);
SELECT create_distributed_table('a3.test1', 'id');

ERROR:  must be member of role "postgres"
CONTEXT:  while executing command on finpg-rcdn-prd-05:5432
```

**Comment by mtuncer** [comment]
Underlying reason is that, we send `create schema xxxx if not exists` command prior to each shard creation.   We piggyback original schema owner information to command thus it becomes
```sql
CREATE SCHEMA IF NOT EXISTS a3 AUTHORIZATION postgres
```

This was used to make sure that the schema owner in coordinator and workers is the same. 

Quick fix is to send create schema command as the same user logged into coordinator. I really like to avoid that if possible.

**Comment by marcocitus** [comment]
Should we connect as the schema owner or admin user when doing the CREATE SCHEMA?

**Comment by mtuncer** [comment]
@marcocitus can we change the connected user on the fly without breaking ongoing transaction ?

**Comment by marcocitus** [comment]
We generally create schemas over a separate connection outside of the distributed transaction, prior to creating shards. Otherwise we'd run into all sorts of deadlocks. I think the reason we still include CREATE SCHEMA in the list of commands to create a shard is to support master_move_shard_placement. It'd be nice to stop doing that.

**Comment by mtuncer** [comment]
I found following workaround to be working on my local cluster

Idea is to create a  *master* role who would own the schema, and add each authorized user to that role group. In following example, I created role `ax` on coordinator and workers, transferred schema ownership to that role.  Then, I created 2 users `a3` and `a2`, and added to group `ax`

I was able to create tables within `a3` schema and distributed them with both users afterwards.

```sql
-- create master role and users
create role ax;
SELECT run_command_on_workers($$create role ax$$);

create role a3 with login encrypted password 'abcd123';
SELECT run_command_on_workers($cmd$create role a3 with login encrypted password 'abcd123';$cmd$);

create role a2 with login encrypted password 'abcd123';
SELECT run_command_on_workers($cmd$create role a2 with login encrypted password 'abcd123';$cmd$);

-- add users to role ax
GRANT ax to a3;
SELECT run_command_on_workers($$GRANT ax to a3$$);

GRANT ax to a2;
SELECT run_command_on_workers($$GRANT ax to a2$$);

grant connect on database fpaprd to a3;
SELECT run_command_on_workers($cmd$grant connect on database fpaprd to a3;$cmd$);

grant connect on database fpaprd to a2;
SELECT run_command_on_workers($cmd$grant connect on database fpaprd to a2;$cmd$);

-- create schema on coordinator and workers
create schema a3;
SELECT run_command_on_workers($cmd$create schema a3;$cmd$);

-- grant privileges to role ax not to individual users
grant create on database fpaprd to ax;
SELECT run_command_on_workers($cmd$grant create on database fpaprd to ax;$cmd$);

grant usage on schema a3 to ax;
SELECT run_command_on_workers($cmd$grant usage on schema a3 to ax;$cmd$);

grant ALL PRIVILEGES on SCHEMA a3 to ax;
SELECT run_command_on_workers($cmd$grant ALL PRIVILEGES on SCHEMA a3 to ax;$cmd$);

grant create on schema a3 to ax;
SELECT run_command_on_workers($cmd$grant create on schema a3 to ax;$cmd$);
```

**Comment by mtuncer** [comment]
@metdos @marcocitus I am thinking about a fix to eliminate remove `create schema if not exists` from shard creation command. Should be able to drop this if we do following
- propagate schema creation commands to workers along with all grants
- add schemas and grants to new node upon completion of master_add_node and its friends

First one should be easy and would solve most problems. 

However,  the second one is not. It looks like it should be part of type/udf propagation work we discussed some time back.

In order for `master_move_shard_placement` to work safely, we need to make sure schema is there (with appropriate privileges set). The schema will be there except when we are targeting a new node.

**Comment by marcocitus** [comment]
Alternatively, we could do a kind of EnsureSchemaExistsOnNode prior to the shard move in master_move_shard_placement. We also do that in create_distributed_table (and then redundantly send it again when creating shards).

**Comment by mtuncer** [comment]
We still have the same problem, `EnsureSchemaExistsOnAllNodes` calls `CreateSchemaDDLCommand` which created `create schema <schema_name> if not exists with authorization <schema owner>`

We this command would fail if the calling user does not have right to create schema in the database.

This command should be send with superuser

**Comment by mtuncer** [comment]
or `CitusExtensionOwnerName()`

**Comment by mtuncer** [comment]
looks like problem went away but we still need to figure out
- master_create_emtpy_shard (append and range tables)
- master_add_node (reference tables)

**Comment by hanefi** [comment]
fixed by #2777
